### PR TITLE
Fixing builds on Windows

### PR DESCRIPTION
**Pull request by dmatej**
I would not recommend MS Windows for a development to anyone, however ... I would like to see the build passing at least basically. Also tests are useful to see differences between operating systems and can suggest some solutions of possible issues.

### CODE REVIEW

**Comment by avpinchuk** [comment]
To save your time, part 1:

https://github.com/eclipse-ee4j/glassfish/blob/67d5fae69e318d008b05741384e454f6dae1c1a0/nucleus/admin/launcher/src/test/java/com/sun/enterprise/admin/launcher/GFLauncherTest.java#L94

https://github.com/eclipse-ee4j/glassfish/blob/67d5fae69e318d008b05741384e454f6dae1c1a0/nucleus/admin/launcher/src/test/java/com/sun/enterprise/admin/launcher/GFLauncherTest.java#L106

`endsWith("java")` works on UNIX but failed on Windows due to `java.exe`.

Also we should disable with `@DisabledOnOs(OS.WINDOWS)` for whole test class:

https://github.com/eclipse-ee4j/glassfish/blob/67d5fae69e318d008b05741384e454f6dae1c1a0/appserver/libpam4j/src/test/java/org/jvnet/libpam/UnixUserTest.java#L26

Asterisk `*` improperly quoted on Windows and `asadmin get *` command fails:

https://github.com/eclipse-ee4j/glassfish/blob/67d5fae69e318d008b05741384e454f6dae1c1a0/appserver/itest-tools/src/test/java/org/glassfish/main/admin/test/tool/StartStopITest.java#L39

**Comment by dmatej** [comment]
> To save your time, part 1:
> 
> https://github.com/eclipse-ee4j/glassfish/blob/67d5fae69e318d008b05741384e454f6dae1c1a0/nucleus/admin/launcher/src/test/java/com/sun/enterprise/admin/launcher/GFLauncherTest.java#L94
> 
> https://github.com/eclipse-ee4j/glassfish/blob/67d5fae69e318d008b05741384e454f6dae1c1a0/nucleus/admin/launcher/src/test/java/com/sun/enterprise/admin/launcher/GFLauncherTest.java#L106
> 
> `endsWith("java")` works on UNIX but failed on Windows due to `java.exe`.
> 
> Also we should disable with `@DisabledOnOs(OS.WINDOWS)` for whole test class:
> 
> https://github.com/eclipse-ee4j/glassfish/blob/67d5fae69e318d008b05741384e454f6dae1c1a0/appserver/libpam4j/src/test/java/org/jvnet/libpam/UnixUserTest.java#L26
> 
> Asterisk `*` improperly quoted on Windows and `asadmin get *` command fails:
> 
> https://github.com/eclipse-ee4j/glassfish/blob/67d5fae69e318d008b05741384e454f6dae1c1a0/appserver/itest-tools/src/test/java/org/glassfish/main/admin/test/tool/StartStopITest.java#L39

Bit late, I already did that, but at least we have same ideas :-)
I ended with this exception in server.log, which is probably caused by some security software not under my control. So I will just push what I have for now ...

```
[2023-09-25T21:52:29.061483+02:00] [GF 7.0.9-SNAPSHOT] [WARNING] [NCLS-JMX-00007] [jakarta.enterprise.system.jmx] [tid: _ThreadID=82 _ThreadName=Thread-9] [levelValue: 900] [[
  Cannot start JMX connector JmxConnector config: { name = system, Protocol = rmi_jrmp, Address = 0.0.0.0, Port = 8686, AcceptAll = false, AuthRealmName = admin-realm, SecurityEnabled = false} due to exception java.io.IOException: Cannot bind to URL [rmi://xxxxx:8686/jmxrmi]: javax.naming.ServiceUnavailableException [Root exception is java.rmi.ConnectException: Connection refused to host:xxxxx; nested exception is: 
	java.net.ConnectException: Connection timed out: connect]]]
```

I cannot make the host name public, but it is from some managed windows domain and I have no rights to expose port on this machine.

**Comment by avpinchuk** [comment]
I added two more issues.

And there are two more tests failed:
[appserver/tests/application/src/test/java/org/glassfish/main/test/app/web/jsp/JspReloadGeneratedServletIfUpdatedTest.java](https://github.com/eclipse-ee4j/glassfish/blob/master/appserver/tests/application/src/test/java/org/glassfish/main/test/app/web/jsp/JspReloadGeneratedServletIfUpdatedTest.java)
https://github.com/eclipse-ee4j/glassfish/blob/master/appserver/tests/embedded/basic/lifecycle/src/test/java/org/glassfish/tests/embedded/basic/lifecycle/LifeCycleTest.java

File: appserver/itest-tools/src/test/java/org/glassfish/main/admin/test/tool/StartStopITest.java [hunk 1]
```diff
@@ -36,6 +36,6 @@ class StartStopITest {
      */
     @Test
     void asadminGet() {
-        assertThat(ASADMIN.exec("get", "*"), asadminOK());
+        assertThat(ASADMIN.exec("get", "'*'"), asadminOK());
```
Comment (avpinchuk) [inline appserver/itest-tools/src/test/java/org/glassfish/main/admin/test/tool/StartStopITest.java:39]: On Windows work with double quotes, on Linux - no. We need some condition here.

Comment (dmatej) [reply appserver/itest-tools/src/test/java/org/glassfish/main/admin/test/tool/StartStopITest.java:39]: Aha, I thought it passed on my Windows, but this test failed to start the domain (unable to expose ports). Can you finish the PR? I can't do more on this machine. I just used the opportunity to try it :)

Comment (avpinchuk) [reply appserver/itest-tools/src/test/java/org/glassfish/main/admin/test/tool/StartStopITest.java]: Sure ;)

**Comment by dmatej** [comment]
Aha, I did not notice you switched it back to Draft, @avpinchuk

**Comment by avpinchuk** [comment]
Yes. This is because we need to close several bugs in GF, not only in tests ;)

**Comment by dmatej** [comment]
> Yes. This is because we need to close several bugs in GF, not only in tests ;)

Ok, I will leave it on your decision for now, thank you.

**Comment by avpinchuk** [comment]
Patch #24611 fixes resource disposition on Windows platform. Now it works:

https://github.com/eclipse-ee4j/glassfish/blob/95e7c0f27dce9ebf044a7881db09fbba6d495b3a/appserver/tests/embedded/basic/lifecycle/src/test/java/org/glassfish/tests/embedded/basic/lifecycle/LifeCycleTest.java#L87

**Comment by avpinchuk** [comment]
On Windows, `Files.copy(src, dst, attrs)` set actual `creationTime` and `lastAcccessTime` for `dst`, but copy `lastModifiedTime` from `src`.

Thus, `JspReloadGeneratedServletIfUpdatedTest` failed because `JspReloadGeneratedServletIfUpdated_jsp` compiled at build time and has earlier `lastModifiedTime` than generated by JSP engine. When we copying build-time version instead of generated, its modified timestamp left unchanged.

**Comment by avpinchuk** [comment]
**Windows 10 Pro 22H2 (19045.3448):**

`mvn clean install -Pstaging`

image

**Comment by avpinchuk** [comment]
@dmatej, all done.

**Comment by dmatej** [comment]
I have yet one failure, the exec plugin execution (and same previous gfbuild-maven-plugin:exec) were depending on OS java command, same as #24617 , I will fix it now.

**Comment by avpinchuk** [comment]
I opened yet another PR to WaSP eclipse-ee4j/wasp#61

**Comment by avpinchuk** [comment]
I started the build...

**Comment by dmatej** [comment]
> I started the build... BUILD SUCCESS

So, let's make it Ready for review, do you agree?

**Comment by avpinchuk** [comment]
> So, let's make it Ready for review, do you agree?

Yes, sure.

**Review by arjantijms — APPROVED** [review]

**Comment by arjantijms** [comment]
LGTM, although I didn't test on Windows. The WaSP PR has some good comments btw ;)

**Review by avpinchuk — APPROVED** [review]

**Comment by avpinchuk** [comment]
> The WaSP PR has some good comments btw ;)

I pushed a fixed version

**Pull request merged**